### PR TITLE
Implement API and data job for last 30 days of PyPI installs

### PR DESCRIPTION
**Pull request by manasaV3**
Relates to: https://github.com/chanzuckerberg/napari-hub/issues/709

#### Happy path - Valid plugin name: 
https://6vwmwypt38.execute-api.us-west-2.amazonaws.com/dev-last30daysInstall/activity/affinder/recentStats

#### Invalid plugin name:
https://6vwmwypt38.execute-api.us-west-2.amazonaws.com/dev-last30daysInstall/activity/test/recentStats

Relates to: 
https://github.com/chanzuckerberg/napari-hub/issues/720

#### Happy path - Valid plugin name: 
https://6vwmwypt38.execute-api.us-west-2.amazonaws.com/dev-last30daysInstall/metrics/affinder

#### Invalid plugin name:
https://6vwmwypt38.execute-api.us-west-2.amazonaws.com/dev-last30daysInstall/metrics/test

### CODE REVIEW

File: backend/api/model.py [hunk 1]
```diff
@@ -92,6 +94,7 @@ def get_frontend_manifest_metadata(plugin, version):
     interpreted_metadata = parse_manifest(raw_metadata)
     return interpreted_metadata
 
+
```
Comment (klai95) [inline backend/api/model.py:97]: Additional newline seems to have been inserted

File: backend/api/model.py [hunk 2]
```diff
@@ -121,7 +125,7 @@ def get_manifest(plugin: str, version: str = None) -> dict:
     elif version is None:
         version = plugins[plugin]
     plugin_metadata = get_cache(f'cache/{plugin}/{version}-manifest.json')
-    
+
```
Comment (klai95) [inline backend/api/model.py:128]: I think you may need to revert the change here too

File: backend/api/model.py [hunk 3]
```diff
@@ -108,6 +111,7 @@ def discover_manifest(plugin: str, version: str = None):
         Payload=json.dumps(lambda_event),
     )
 
+
```
Comment (klai95) [inline backend/api/model.py:114]: Additional newline here too

File: backend/api/model.py [hunk 4]
```diff
@@ -377,52 +381,44 @@ def update_activity_data():
...
+        SELECT 
+            file_project, DATE_TRUNC('month', timestamp) as month, count(*) as num_downloads
+        FROM
+            imaging.pypi.labeled_downloads
+        WHERE 
+            download_type = 'pip'
+            AND project_type = 'plugin'
+        GROUP BY file_project, month
+        ORDER BY file_project, month
+        """
```
Comment (klai95) [inline backend/api/model.py:394]: nit: should the bottom triple quote align with query? Just want to make sure

File: backend/api/model.py [hunk 5]
```diff
@@ -377,52 +381,44 @@ def update_activity_data():
...
+            file_project, count(*) as num_downloads
+        FROM
+            imaging.pypi.labeled_downloads
+        WHERE 
+            download_type = 'pip'
+            AND project_type = 'plugin'
+            AND timestamp > DATEADD(DAY, -30, CURRENT_DATE)
+        GROUP BY file_project     
+        ORDER BY file_project
+        """
```
Comment (klai95) [inline backend/api/model.py:419]: same question as above

File: backend/api/model.py [hunk 6]
```diff
@@ -458,3 +454,41 @@ def get_installs_stats(plugin: str) -> Any:
...
     return obj
+
+def get_recent_installs_stats(plugin: str) -> Any:
+    """
+    This should return a dict, with number of Installs in the last 30 days
+
+    :param plugin: plugin name
+    :return: dict
+    """
+    response = {'installInLast30Days': get_recent_activity_dashboard_data().get(plugin, 0)}
```
Comment (richaagarwal) [inline backend/api/model.py:465]: This is very minor, but I think the plural `installs` would make more sense in the naming here, so `installsInLast30Days`.

**Comment by richaagarwal** [comment]
I'd suggest giving the pull request a more readable name, e.g. "Implement API endpoint for last 30 days of PyPI installs"

File: backend/api/model.py (hunk 1 above)
Comment (manasaV3) [reply backend/api/model.py:99]: I added the new line to be consistent with recommendations from [here](https://peps.python.org/pep-0008/#blank-lines) and also most of the other methods had two lines separating them.
Lemme know, if we are following a different formatting. 😅

File: backend/api/model.py (hunk 4 above)
Comment (manasaV3) [reply backend/api/model.py]: Yeah, there should be indent on the query.

File: backend/api/model.py (hunk 5 above)
Comment (manasaV3) [reply backend/api/model.py]: Yeah, there should be indent on the query.

File: backend/api/dao/s3.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import os
+from typing import Dict, Any
+
+import boto3
+from botocore.exceptions import ClientError
+
+from utils.timer import Timer
+
+
+class S3DAO:
```
Comment (richaagarwal) [inline backend/api/dao/s3.py:12]: I don't know too much about the Data Access Object pattern in Java specifically, but I don't think DAO needs to be a part of the naming here.  Something more descriptive might be `HubS3` if we want to define our own set of functions to interface with boto. That said, I could imagine it would get a little redundant if we had to provide a wrapper method here each time we want to call a new boto utility. 

I think we can improve the functions available in https://github.com/chanzuckerberg/napari-hub/blob/main/backend/api/s3.py such that they are more reusable and enable easier logging, but am curious about the need for our own class.

File: backend/api/dao/s3.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    def __init__(self):
+        self.__bucket = os.environ.get('BUCKET')
+        self.__bucket_path = os.environ.get('BUCKET_PATH', '')
+        __endpoint_url = os.environ.get('BOTO_ENDPOINT_URL', None)
+        self.__s3_client = boto3.client('s3', endpoint_url=__endpoint_url)
+
+    def get_activity_timeline_data(self) -> Any:
+        try:
+            return self.__get_object_body_from_s3('activity_dashboard_data/plugin_installs.csv').decode('utf-8')
+        except Exception as e:
```
Comment (richaagarwal) [inline backend/api/dao/s3.py:23]: It can be useful to catch particular exceptions we expect, and `raise` those we don't. I would caution against only logging here.

File: backend/api/dao/s3.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            logging.error(e)
+            return None
+
+    def get_recent_activity_dashboard_data(self) -> Dict:
+        path = 'activity_dashboard_data/plugin_recent_installs.json'
+        try:
```
Comment (richaagarwal) [inline backend/api/dao/s3.py:29]: For error handling/logging, this could probably be captured in a decorator function, that you can then apply to the desired functions, to avoid redundant code. Lots of resources on decorators in python -- https://higee.io/exception-handling-decorators-with-proper-logging-d0f5b54f7ba8 has some examples specifically for exception handling.

File: backend/api/dao/s3.py (hunk 7 above)
Comment (manasaV3) [reply backend/api/dao/s3.py:12]: I wanted to have something that was following Single Responsibility Principle, where the only responsibility would be to interface with s3 and have no additional logic to it. 

It was easier to implement that for the activity pieces as the logic was easier to cleanly separate out. I wanted to clean up the other methods in api.s3 with more thought, so I left them as is in that module albeit the timer. :)

File: backend/api/dao/s3.py (hunk 9 above)
Comment (manasaV3) [reply backend/api/dao/s3.py:29]: I was planning on making the timer a decorator in the near future, but I am happy to do that here along with the logging. 😁

File: backend/api/s3.py [hunk 10]
```diff
@@ -82,3 +82,11 @@ def get_activity_dashboard_data(plugin) -> Dict:
     plugin_df['MONTH'] = pd.to_datetime(plugin_df['MONTH'])
     return plugin_df
 
+
+def get_recent_activity_dashboard_data():
```
[the remark this answers is not in the document]
Comment (manasaV3) [reply backend/api/s3.py]: I was planning on retaining that logic in the model, to support the SRP. What are your thoughts on that?

[the remark this answers is not in the document]
Comment (richaagarwal) [reply backend/api/s3.py]: That sounds good to me!

File: backend/api/dao/s3.py (hunk 9 above)
Comment (richaagarwal) [reply backend/api/dao/s3.py:29]: Sounds good - doesn't have to be in this PR!

File: backend/api/dao/snowflake_db.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        SELECT 
+            file_project, DATE_TRUNC('month', timestamp) as month, count(*) as num_downloads
+        FROM
+            imaging.pypi.labeled_downloads
+        WHERE 
+            download_type = 'pip'
+            AND project_type = 'plugin'
+        GROUP BY file_project, month
+        ORDER BY file_project, month
+    """
```
Comment (klai95) [inline backend/api/dao/snowflake_db.py:20]: Should there be a newline between the two queries?

File: backend/api/app.py [hunk 12]
```diff
@@ -126,6 +127,16 @@ def get_plugin_installs_stats(plugin: str) -> Response:
     return jsonify(get_installs_stats(plugin))
 
 
+@app.route('/activity/<plugin>/recent_stats')
+def get_recent_plugin_installs_stats(plugin: str) -> Response:
+    return jsonify(get_recent_installs_stats(plugin))
+
+
+@app.route('/metrics/<plugin>')
```
Comment (klai95) [inline backend/api/app.py:135]: Just to get more context, what are some examples of `<metrics>`?

**Comment by manasaV3** [comment]
Closing this in favour of https://github.com/chanzuckerberg/napari-hub/pull/741
